File: data/java/src/com/db/data/format/AviStreamHeaderList.java
/*
 * Copyright (c) 2007 Digital Bazaar, Inc.  All rights reserved.
 */
package com.db.data.format;

import java.io.IOException;
import java.io.OutputStream;

/**
 * An AVI Stream Header List ('strl').
 * 
 * AVI Format is as follows:
 * 
 * AVI Form Header ('RIFF' size 'AVI ' data)
 *    Header List ('LIST' size 'hdrl' data)
 *       AVI Header ('avih' size data)
 *       Video Stream Header List ('LIST' size 'strl' data)
 *          Video Stream Header ('strh' size data)
 *          Video Stream Format ('strf' size data)
 *          Video Stream Data ('strd' size data) - IGNORED, for DRIVERS
 *       Audio Stream Header List ('LIST' size 'strl' data)
 *          Audio Stream Header ('strh' size data)
 *          Audio Stream Format ('strf' size data)
 *          Audio Stream Data ('strd' size data) - IGNORED, for DRIVERS
 *    Info List ('LIST' size 'INFO' data)
 *       Index Entry ({'ISBJ','IART','ICMT',...} size data)
 *    Movie List ('LIST' size 'movi' data)
 *       Movie Entry ({'00db','00dc','01wb'} size data)
 *    Index Chunk ('idx1' size data)
 *       Index Entry ({'00db','00dc','01wb',...})
 * 
 * The Stream Header List 'strl' applies to the first stream in the 'movi'
 * LIST, the second applies to the second stream, etc.
 * 
 * @author Dave Longley
 */
public class AviStreamHeaderList
{
   /**
    * The AVI stream header RIFF header.
    */
   protected RiffListHeader mRiffHeader;
   
   /**
    * The Stream Header.
    */
   protected AviStreamHeader mStreamHeader;
   
   /**
    * The Stream Format.
    */
   protected AviStreamFormat mStreamFormat;
   
   /**
    * The Stream Data.
    */
   protected AviStreamData mStreamData;
   
   /**
    * Constructs a new AviStreamHeaderList.
    */
   public AviStreamHeaderList()
   {
      // create RIFF header
      mRiffHeader = new RiffListHeader("strl");
      
      // create stream header and format
      mStreamHeader = new AviStreamHeader();
      mStreamFormat = new AviStreamFormat();
      
      // no stream data by default
      mStreamData = null;
   }
   
   /**
    * Writes this AviStreamHeaderList, including the RIFF header, to an
    * OutputStream.
    * 
    * @param os the OutputStream to write to.
    * 
    * @exception IOException thrown if an IO error occurs.
    */
   public void writeTo(OutputStream os) throws IOException
   {
      // write RIFF header
      mRiffHeader.writeTo(os);
      
      // write stream header
      mStreamHeader.writeTo(os);
      
      // write stream format
      mStreamFormat.writeTo(os);
      
      // write stream data, if present
      if(mStreamData != null)
      {
         mStreamData.writeTo(os);
      }
   }
   
   /**
    * Converts this AviStreamHeaderList from a byte array.
    * 
    * @param b the byte array to convert from.
    * @param offset the offset to start converting from.
    * @param length the number of valid bytes in the buffer following the
    *               offset.
    * 
    * @return true if successful, false if not.
    */
   public boolean convertFromBytes(byte[] b, int offset, int length)
   {
      boolean rval = false;
      
      // convert the RIFF header
      if(mRiffHeader.convertFromBytes(b, offset, length) &&
         mRiffHeader.getIdentifier().equals("strl"))
      {
         // step forward past RIFF header
         offset += RiffListHeader.LIST_HEADER_SIZE;
         length -= RiffListHeader.LIST_HEADER_SIZE;
         
         // make sure there is enough data to convert the header
         if(length >= mRiffHeader.getListSize())
         {
            // convert AviStreamHeaders:
            
            // FIXME:
            
            
            // converted successfully
            rval = true;
         }
      }
      
      return rval;
   }
   
   /**
    * Returns whether or not this AviStreamHeaderList is valid.
    * 
    * @return true if valid, false if not.
    */
   public boolean isValid()
   {
      return mRiffHeader.isValid();
   }
   
   /**
    * Gets the size of this AviStreamHeaderList, not including its chunk header.
    * 
    * @return the size of this AviStreamHeaderList.
    */
   public long getSize()
   {
      return mRiffHeader.getListSize();
   }
}

File: data/cpp/db/data/avi/AviStreamHeaderList.cpp
/*
 * Copyright (c) 2007-2008 Digital Bazaar, Inc.  All rights reserved.
 */

#include "db/data/avi/AviStreamHeaderList.h"

using namespace db::data;
using namespace db::data::avi;
using namespace db::data::riff;
using namespace db::io;

AviStreamHeaderList::AviStreamHeaderList() :
   mRiffHeader(CHUNK_ID)
{
}

AviStreamHeaderList::~AviStreamHeaderList()
{
}

bool AviStreamHeaderList::writeTo(OutputStream& os)
{
   bool rval;
   
   // write RIFF header
   rval = mRiffHeader.writeTo(os);
   
   if(rval)
   {
      // write stream header
      rval = mStreamHeader.writeTo(os);
   }
   
   if(rval)
   {
      // write stream format
      rval = mStreamFormat.writeTo(os);
   }
   
   if(rval)
   {
      // write stream data, if present
      if(mStreamData.getChunkSize() > 0)
      {
         rval = mStreamData.writeTo(os);
      }
   }
   
   return rval;
}

bool AviStreamHeaderList::convertFromBytes(const char* b, int offset, int length)
{
   bool rval = false;
   
   // convert the RIFF header
   if(mRiffHeader.convertFromBytes(b, offset, length) &&
      mRiffHeader.getIdentifier() == CHUNK_ID)
   {
      // make sure there is enough data to convert the header
      if(length >= getSize())
      {
         // step forward past RIFF header
         offset += RiffListHeader::HEADER_SIZE;
         
         // set length to list size
         length = (int)mRiffHeader.getListSize();
         
         // convert header
         if(mStreamHeader.convertFromBytes(b, offset, length))
         {
            // step forward past header
            offset += mStreamHeader.getSize();
            length -= mStreamHeader.getSize();
            
            // convert format
            if(mStreamFormat.convertFromBytes(b, offset, length))
            {
               // header list converted, 'strd' is not used
               rval = true;
               
               // step forward past format
               offset += mStreamFormat.getSize();
               length -= mStreamFormat.getSize();
               
               // look for stream data anyway
               if(length > 0)
               {
                  // convert stream data
                  mStreamData.convertFromBytes(b, offset, length);
               }
            }
         }
      }
   }
   
   return rval;
}

bool AviStreamHeaderList::isValid()
{
   return mRiffHeader.isValid() && mRiffHeader.getIdentifier() == CHUNK_ID;
}

int AviStreamHeaderList::getListSize()
{
   // AVI stream header list is expected to be much smaller than 32-bits
   return (int)mRiffHeader.getListSize();
}

int AviStreamHeaderList::getSize()
{
   return getListSize() + RiffListHeader::HEADER_SIZE;
}
